// @elevenlabs/client/dist/platform/VoiceSessionSetup.js
import { WebRTCConnection } from "../utils/WebRTCConnection.js";
/**
 * The active session setup strategy.
 * Defaults to undefined — set by platform-specific entrypoints on import.
 */
export let setupStrategy;
/**
 * Override the voice session setup strategy.
 * Called by platform-specific entrypoints to inject their own setup handling.
 */
export function setSetupStrategy(strategy) {
    setupStrategy = strategy;
}
/**
 * Sets up a voice session for a WebRTC connection.
 * Platform-agnostic: extracts the input/output controllers that
 * WebRTCConnection already provides via LiveKit.
 */
export function setupWebRTCSession(connection) {
    if (!(connection instanceof WebRTCConnection)) {
        throw new Error("setupWebRTCSession requires a WebRTCConnection. " +
            `Received: ${connection?.constructor?.name ?? typeof connection}`);
    }
    return {
        connection,
        input: connection.input,
        output: connection.output,
        playbackEventTarget: null,
        detach: async () => { },
    };
}
