// @elevenlabs/client/dist/index.js
import { isTextOnly } from "./BaseConversation.js";
import { assertRuntimeCompatibility } from "./runtime.js";
import { TextConversation } from "./TextConversation.js";
import { VoiceConversation } from "./VoiceConversation.js";
export { DEFAULT_INPUT_CHUNK_DURATION_MS } from "./InputController.js";
export { createConnection } from "./utils/ConnectionFactory.js";
export { WebSocketConnection } from "./utils/WebSocketConnection.js";
export { WebRTCConnection } from "./utils/WebRTCConnection.js";
export { postOverallFeedback } from "./utils/postOverallFeedback.js";
export { SessionConnectionError } from "./utils/errors.js";
export { VoiceConversation } from "./VoiceConversation.js";
export { TextConversation } from "./TextConversation.js";
// Scribe exports
export { Scribe, AudioFormat, CommitStrategy, RealtimeEvents, RealtimeConnection, } from "./scribe/index.js";
export const Conversation = {
    async startSession(options) {
        assertRuntimeCompatibility();
        return isTextOnly(options)
            ? TextConversation.startSession(options)
            : VoiceConversation.startSession(options);
    },
};
