// @elevenlabs/client/dist/VoiceConversation.js
import { BaseConversation, } from "./BaseConversation.js";
import { setupStrategy } from "./platform/VoiceSessionSetup.js";
export class VoiceConversation extends BaseConversation {
    input;
    output;
    playbackEventTarget;
    cleanUp;
    type = "voice";
    static async startSession(options) {
        const fullOptions = BaseConversation.getFullOptions(options);
        if (fullOptions.onStatusChange) {
            fullOptions.onStatusChange({ status: "connecting" });
        }
        if (fullOptions.onCanSendFeedbackChange) {
            fullOptions.onCanSendFeedbackChange({ canSendFeedback: false });
        }
        let conversation = null;
        let sessionSetup = null;
        try {
            // Platform-specific strategy handles wake lock, mic permission,
            // delay, connection creation, and input/output setup.
            if (!setupStrategy) {
                throw new Error("No voice session setup strategy registered. " +
                    'Import the platform-specific entry point (e.g. @elevenlabs/client via the "browser" export).');
            }
            sessionSetup = await setupStrategy(fullOptions);
            conversation = new VoiceConversation(fullOptions, sessionSetup.connection, sessionSetup.input, sessionSetup.output, sessionSetup.playbackEventTarget, sessionSetup.detach);
            fullOptions.onConversationCreated?.(conversation);
            conversation.markConnected();
            fullOptions.onConnect?.({
                conversationId: sessionSetup.connection.conversationId,
            });
            return conversation;
        }
        catch (error) {
            if (conversation) {
                await conversation.endSession().catch(() => { });
            }
            else {
                // Strategy returned but conversation wasn't created — clean up
                if (sessionSetup) {
                    await sessionSetup.detach().catch(() => { });
                }
                fullOptions.onStatusChange?.({ status: "disconnected" });
            }
            throw error;
        }
    }
    inputFrequencyData;
    outputFrequencyData;
    handlePlaybackEvent = event => {
        if (event.data.type === "process") {
            this.updateMode(event.data.finished ? "listening" : "speaking");
        }
    };
    constructor(options, connection, input, output, playbackEventTarget, cleanUp) {
        super(options, connection);
        this.input = input;
        this.output = output;
        this.playbackEventTarget = playbackEventTarget;
        this.cleanUp = cleanUp;
        playbackEventTarget?.addListener(this.handlePlaybackEvent);
    }
    async handleEndSession() {
        this.playbackEventTarget?.removeListener(this.handlePlaybackEvent);
        this.playbackEventTarget = null;
        await this.cleanUp();
        await super.handleEndSession();
        await this.input.close();
        await this.output.close();
    }
    handleInterruption(event) {
        super.handleInterruption(event);
        this.updateMode("listening");
        this.output.interrupt();
    }
    handleAudio(event) {
        super.handleAudio(event);
        if (event.audio_event.alignment && this.options.onAudioAlignment) {
            this.options.onAudioAlignment(event.audio_event.alignment);
        }
        if (this.lastInterruptTimestamp <= event.audio_event.event_id) {
            if (event.audio_event.audio_base_64) {
                this.options.onAudio?.(event.audio_event.audio_base_64);
                // Audio routing is handled by attachConnectionToOutput for WebSocket
                // WebRTC handles audio playback directly through LiveKit tracks
            }
            this.currentEventId = event.audio_event.event_id;
            this.updateCanSendFeedback();
            this.updateMode("speaking");
        }
    }
    static FREQUENCY_BIN_COUNT = 1024;
    setMicMuted(isMuted) {
        this.input.setMuted(isMuted).catch(error => {
            this.options.onError?.("Failed to set input muted state", error);
        });
    }
    getInputByteFrequencyData() {
        this.inputFrequencyData ??= new Uint8Array(VoiceConversation.FREQUENCY_BIN_COUNT);
        this.input.getByteFrequencyData(this.inputFrequencyData);
        return this.inputFrequencyData;
    }
    getOutputByteFrequencyData() {
        this.outputFrequencyData ??= new Uint8Array(VoiceConversation.FREQUENCY_BIN_COUNT);
        this.output.getByteFrequencyData(this.outputFrequencyData);
        return this.outputFrequencyData;
    }
    getInputVolume() {
        return this.input.getVolume();
    }
    getOutputVolume() {
        return this.output.getVolume();
    }
    async changeInputDevice({ sampleRate, format, preferHeadphonesForIosDevices, inputDeviceId, }) {
        try {
            await this.input.setDevice({
                inputDeviceId,
                sampleRate,
                format,
                preferHeadphonesForIosDevices,
            });
        }
        catch (error) {
            console.error("Error changing input device", error);
            throw error;
        }
    }
    async changeOutputDevice({ sampleRate, format, outputDeviceId, }) {
        try {
            await this.output.setDevice({
                outputDeviceId,
                sampleRate,
                format,
            });
        }
        catch (error) {
            console.error("Error changing output device", error);
            throw error;
        }
    }
    setVolume = ({ volume }) => {
        // clamp & coerce
        const clampedVolume = Number.isFinite(volume)
            ? Math.min(1, Math.max(0, volume))
            : 1;
        this.volume = clampedVolume;
        // Delegate to output controller
        this.output.setVolume(clampedVolume);
    };
}
